### ISSUE TITLE
RecordWildCards should support the "empty record"

### ISSUE THREAD

**Issue by Ericson2314** [ghc-proposals/ghc-proposals#484]
Forked out from https://github.com/haskell/error-messages/issues/15

```haskell
data Foo = Foo { a :: Int }

f (Foo {..}) = 1
```
being allowed, but
```haskell
data Foo = Foo { }

f (Foo {..}) = 1
```
not being allowed, feels arbitrary and annoying. The empty field has a name for every field, so on what grounds do we say it is not the empty record?

-----------------------

Note that I had thought Rust did this, allowing all ways of pattern matching on an empty `struct` regardless of how it was defined, but I am in fact mistaken:  
```rust
struct Foo();
struct Bar {}
struct Baz;

fn main() {

  //// illegal because shadowing (?)
  //match (Foo {}) {
  //  Foo => 1, // binds a name Foo, rather than pattern matches on constructor
  //};
  match (Foo {}) {
    Foo() => 1,
  };
  match Foo() {
    Foo {} => 1,
  };
  
  match (Bar {}) {
    Bar => 1, // binds a name Bar, rather than pattern matches on constructor
  };
  match (Bar {}) {
    //Bar() => 1,
    Bar {} => 1,
  };
  //match Bar() {
  match (Bar {}) {
    Bar {} => 1,
  };
  
  match Baz {
    Baz => 1,
  };
  match (Baz{}) {
    //Baz() => 1,
    Baz {} => 1,
  };
  //match Baz() {
  match Baz {
    Baz {} => 1,
  };
}
```

The commented combinations are those that do not work.

I can't figure out the pattern here; I guess it's not such clear precedent. But do note the record patterns are in fact the only sort that works across all 3 definitions.

**Comment by treeowl** [comment]
What possible benefit could this change have?

**Comment by Ericson2314** [comment]
Learning and simplicity. Someone found the error message annoying enough to file and error message. I think that error has no reason to exist -- rather than debate how to rephrase it was can just delete the code that issues it. Any code that turns fields into free bindings will naturally do the right thing when there are no fields.

Arguably,
```haskell
data Foo = Foo
f (Foo {}) = 1
```
in Haskell means we *already* deem fieldless constructors empty records. It is simply more uniform if `{ .. }` works there too.

**Comment by tomjaguarpaw** [comment]
Strong support from me on removing this special case.

**Comment by mpscholten** [comment]
> What possible benefit could this change have?

This error comes up often when using IHP code generators. When e.g. someone generates a view for an action we generate. a file like this:

```haskell
data LogOutView = LogOutView

instance View LogOutView where
    html LogOutView { .. } = [hsx||]
```

The developer is supposed to add some fields to `LogOutView` at some point when needed. We want all view variables to be automatically in scope in the `html` definition, therefore we have the `LogOutView { .. }` pattern there. But until the developer adds a field to `LogOutView`, compiling will fail.

It might sound like a trivial problem, but I've seen many beginners getting stuck here as the error message is not really helpful in these cases.

**Comment by nomeata** [comment]
Thanks for the concrete, practical use case, it makes the proposal much stronger for me.

**Comment by goldfirere** [comment]
I'm in support. Would someone write this up as a proposal? It needn't be very long! But including the concrete motivation would be helpful indeed.

**Comment by simonpj** [comment]
I'm in support too.  A short GHC proposal is the way to go.  I think it'd be accepted promptly.

**Comment by adamgundry** [comment]
The discussion above has been about pattern-matching, but what about using record wildcards in construction of datatypes with no fields, e.g.
```hs
data Foo = Foo
f (Foo {..}) = Foo{..}
```
Presumably this would be allowed too? I think it makes sense to treat records with N labelled fields consistently when N=0. Presumably if there were unlabelled fields (e.g. `data Foo = Foo Int`) the use of record wildcards would be rejected, however?

---

A not entirely related thought: is there a clear reason why record wildcards don't work for updates with at least one field specified, e.g.
```hs
data Bar = MkBar { x :: Int, y :: Int }
g bar y = bar { x = 0, .. }
```
This seems possible to support as resolving the name `x` determines the type `Bar` being updated, and so determines which additional fields are part of the wildcard. That said, I'm not sure we should actually change this now, not least because `OverloadedRecordUpdate` clearly couldn't work with wildcards, and it would suffer from the inconsistency that a record update consisting only of a wildcard such as `bar { .. }` could not be supported.

**Comment by Ericson2314** [comment]
https://github.com/ghc-proposals/ghc-proposals/pull/496 OK finally got around to making a proposal; sorry for the delay.

**Comment by sjakobi** [comment]
Can this issue be closed now that #496 has been merged?
